### PR TITLE
Pubsub Source

### PR DESCRIPTION
**Pull request by nachocano**
Fixes https://github.com/google/knative-gcp/issues/303

* Adding a PubSub source CRD. 
* Adding types, defaults, validation.
* Adding reconciler, which just creates a PullSubscription
* Currently exposing all stuff from PullSubscription but the mode.

* Pending (will sync with @n3wscott personally)
  * need to move PubSub base reconciler to some other folder, otherwise it is somewhat confusing... Can be done in a subsequent PR

/cc @n3wscott @vaikas-google

### CODE REVIEW

File: pkg/reconciler/pubsub/pubsub.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+const (
+	finalizerName = controllerAgentName
+
+	resourceGroup = "pubsubs.events.cloud.run"
+)
+
+// Reconciler is the controller implementation for the PubSub source.
+type Reconciler struct {
+	*reconciler.PubSubBase
```
Comment (n3wscott) [inline pkg/reconciler/pubsub/pubsub.go]: I think we can just have normal base.

File: pkg/reconciler/pubsub/pubsub.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+	if topic == "" {
+		topic = ps.Spec.Topic
+	}
+
+	// See if the source has been deleted.
+	deletionTimestamp := ps.DeletionTimestamp
+
+	if deletionTimestamp != nil {
+		if err := c.PubSubBase.DeletePubSub(ctx, ps.Namespace, ps.Name); err != nil {
```
Comment (n3wscott) [inline pkg/reconciler/pubsub/pubsub.go]: This rec should not worry about talking to punbsub, just delete the PullSubscription.

Comment (n3wscott) [reply pkg/reconciler/pubsub/pubsub.go]: or let the owner refs cleanup

File: pkg/reconciler/pubsub/pubsub.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+			c.Logger.Infof("Unable to delete pubsub resources : %s", err)
+			return fmt.Errorf("failed to delete pubsub resources: %s", err)
+		}
+		c.removeFinalizer(ps)
+		return nil
+	}
+
+	// Ensure that there's finalizer there, since we're about to attempt to
+	// change external state with the topic, so we need to clean it up.
+	err := c.ensureFinalizer(ps)
```
Comment (n3wscott) [inline pkg/reconciler/pubsub/pubsub.go]: This resource likely does not need a finalizer

File: pkg/reconciler/pubsub/pubsub.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	if err != nil {
+		return err
+	}
+
+	t, psubs, err := c.PubSubBase.ReconcilePubSub(ctx, ps, topic, resourceGroup)
```
Comment (n3wscott) [inline pkg/reconciler/pubsub/pubsub.go]: The main role of the PubSub will be to propagate the status of the PullSubscription to this resource

**Comment by nachocano** [comment]
> The resource looks good, I think the reconciler should be much simpler:
> 
> 1. Just make and own a PullSubscription
> 2. Propagate the status of that PullSubscription into the status of the PubSub
> 3. No need to talk to the PubSub API, if the topic does not exist, then that is fine... the pullsubscription will let you know there is an issue.

Sounds good! Will do. Should be done by Monday at most... Sorry for the delay, been entertained with eventing-contrib

**Comment by n3wscott** [comment]
>Sorry for the delay

No need to hurry. :D

File: pkg/reconciler/pubsub/pubsub.go (hunk 1 above)
Comment (nachocano) [reply pkg/reconciler/pubsub/pubsub.go]: Done

File: pkg/reconciler/pubsub/pubsub.go (hunk 2 above)
Comment (nachocano) [reply pkg/reconciler/pubsub/pubsub.go]: done! thanks!

File: pkg/reconciler/pubsub/pubsub.go (hunk 3 above)
Comment (nachocano) [reply pkg/reconciler/pubsub/pubsub.go]: removed, thanks!

File: pkg/reconciler/pubsub/pubsub.go (hunk 4 above)
Comment (nachocano) [reply pkg/reconciler/pubsub/pubsub.go]: removed, thanks!

**Comment by nachocano** [comment]
/hold cancel

@n3wscott ready for another pass

**Pull request merged**